[DOC] Add more exposition to composition example (#2536)
NVIDIA/cutlass: media/docs/cpp/cute/02_layout_algebra.md

With the above, we can assume without loss of generality that `B = s:d` is a layout with integral shape and stride. We can also assume that `A` is a flattened, coalesced layout.

> When `A` is integral, `A = a:b`, the result is rather trivial: `R = A o B = a:b o s:d = s:(b*d)`. Here the composition `R` selects `s` elements, indexed from the natural coordinate 0, strided by `d`.

Reviewer: ```suggestion
When `A` is integral, `A = a:b`, the result is rather trivial: `R = A o B = a:b o s:d = s:(b*d)`. Here, the result of the composition `R` are the first `s` elements of `A` strided by `d`.
```

When `A` is multimodal, we need to be more careful. Put into words, `A o B = A o s:d`, for integral `s` and `d` means that we want to:

> 1. Find a view of `A` such that we incorporate the stride `d` of `B`. This also allows us to to calculate the stride of the composed layout.

Reviewer: ```suggestion
1. Determine a layout that produces every `d`th element of `A`.
```

> The shape of the view can be calculated by progressively "dividing out" the first `d` elements from the shape of `A` starting from the left.

Reviewer: ```suggestion
The shape of this intermediate layout can be calculated by progressively "dividing out" the first `d` elements from the shape of `A` starting from the left.
```

* `(3,6,2,8) /  9 => (1,2,2,8)`
* `(3,6,2,8) / 72 => (1,1,1,4)`

> To understand how this allows us to calculate the stride of the composition, consider for the composition `(6,2):(x, y) o (2, 2)`, where `x, y` are integers.

Reviewer 2: Do you mean:

"... consider the composition `(6, 2): (x, y) o 2:2`?

> To understand how this allows us to calculate the stride of the composition, consider for the composition `(6,2):(x, y) o (2, 2)`, where `x, y` are integers.

Reviewer: ```suggestion
To compute the strides of the strided layout, the residues of the above operation are used to scale the strides of `A`. For instance, the last example `(3,6,2,8):(w,x,y,z) / 72` with strides `(w,x,y,z)` produces `(3*w,6*x,2*x,2*z)` as the strides of the strided layout.
```

> "Dividing out" to find the shape of view we see that `(6,2) / 2 = (3,2)`, implying that the stride of the composed layouts is `(2*x, y)` - incorporating both strides.

Reviewer: ```suggestion
```

As you may have noticed, we can only divide shapes by certain values and get a sensible result. This is called the **stride divisibility condition** and is statically checked in CuTe when possible.

> 2. Keep the first `s` elements of the newly strided `A`. This defines the shape of the composed layout. This can be computed by "modding out" the first `s` elements from the shape of `A` starting from the left.

Reviewer: ```suggestion
2. Keep the first `s` elements of the newly strided `A` so that the result has a compatible shape with `B`. This can be computed by "modding out" the first `s` elements from the shape of `A` starting from the left.
```

Again, this operation must satisfy a **shape divisibility condition** to yield a sensible result and is statically checked in CuTe when possible.

Returning to our example `(6,2):(x, y) o (2, 2)`, we see that "modding out" to find the required shape results in `(6, 2) % 2 = (2, 1)`. This results in a final composition `(6,2):(x, y) o (2, 2) = (2, 1): (2*x, y)`.

---

#### Example 1 -- Worked Example of Calculating a Composition

We provide a more complex example of composition, where both operand layouts are multi-modal to illustrate the concepts introduced above.

```
Functional composition, R := A o B
R(c) := (A o B)(c) := A(B(c))

Example
A = (6,2):(8,2)
B = (4,3):(3,1)

1. Using the left-distributive and concatenation properties for layouts we write the composition as,

R = A o B
  = (6,2):(8,2) o (4,3):(3,1)
  = ((6,2):(8,2) o 4:3, (6,2):(8,2) o 3:1)

---
1. Beginning with (6,2):(8,2) o 4:3

- Calculate the shape of the new strided view of the composition

(6, 2) / 3 = (2, 2)

- This implies a stride of

(8*3, 2) = (24, 2)

- Calculate the shape of the composition

(6, 2) % 4 = (2, 2)

> - We note that the operation above must respect the two-dimensional nature of the stride over A, so (4, 1) is not acceptable.

Reviewer: This is the incorrect intuition and is misleading.

First, a version of `A` that is strided by `d` is computed. Then, the first `s` elements *of that strided layout* are "kept". The first `s` elements of the original `A` play no role and there is no "acceptability" or "two-dimensional" nature to consider.

---

> 2. Now compute (6,2):(8,2) o 3:1
> 
> - Again calculate the shape of the new strided view of the composition
> 
> (6, 2) / 1 = (1, 1) = 1
> 
> - Calculate the shape of the composition
> 
> (6, 2) % 3 = (3, 1) = 3

Reviewer: ```suggestion
2. Compute `(6,2):(8,2) o 3:1`

- First, we compute the strided layout

(6,2):(8,2) / 1 = (6,2):(8,2)

- Next, we keep the shape compatible,

(6,2):(8,2) % 3 = (3,1):(8,2)
```

---

> Putting this together, we see that the composition

Reviewer: ```suggestion
Putting this together and coalescing each mode, we obtain the result
```
